flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java

```diff
@@ -239,25 +294,75 @@ public void run(SourceContext<OUT> ctx) throws Exception {
 		}
 	}
 
-	private class RMQCollector implements Collector<OUT> {
-
+	/**
+	 * Special collector for RMQ messages.
+	 * Captures the correlation ID and delivery tag also does the filtering logic for weather a message has been
+	 * processed or not.
+	 */
+	private class RMQCollectorImpl implements RMQDeserializationSchema.RMQCollector<OUT> {
 		private final SourceContext<OUT> ctx;
 		private boolean endOfStreamSignalled = false;
+		private long deliveryTag;
+		private Boolean preCheckFlag;
 
-		private RMQCollector(SourceContext<OUT> ctx) {
+		private RMQCollectorImpl(SourceContext<OUT> ctx) {
 			this.ctx = ctx;
 		}
 
 		@Override
 		public void collect(OUT record) {
-			if (endOfStreamSignalled || schema.isEndOfStream(record)) {
-				this.endOfStreamSignalled = true;
+			Preconditions.checkNotNull(preCheckFlag, "setCorrelationID must be called at least once before" +
+				"calling this method !");
+
+			if (!preCheckFlag) {
 				return;
 			}
 
+			if (isEndOfStream(record)) {
+				this.endOfStreamSignalled = true;
+				return;
+			}
 			ctx.collect(record);
 		}
 
+		public void collect(List<OUT> records) {
+			Preconditions.checkNotNull(preCheckFlag, "setCorrelationID must be called at least once before" +
+				"calling this method !");
+
+			if (!preCheckFlag) {
+				return;
+			}
+
+			for (OUT record : records){
+				if (isEndOfStream(record)) {
+					this.endOfStreamSignalled = true;
+					return;
+				}
+				ctx.collect(record);
+			}
+		}
+
+		public void setMessageIdentifiers(String correlationId, long deliveryTag){
+			preCheckFlag = true;
+			if (!autoAck) {
+				if (usesCorrelationId) {
+					Preconditions.checkNotNull(correlationId, "RabbitMQ source was instantiated " +
+						"with usesCorrelationId set to true yet we couldn't extract the correlation id from it !");
+					if (!addId(correlationId)) {
+						// we have already processed this message
+						preCheckFlag = false;
+					}
+				}
+				sessionIds.add(deliveryTag);
+			}
+		}
```

Comment:
What do you think about keeping this as an internal method that just sets the current message's corrleationId and tag? And we could call it from the `collect(String correlationId, long deliveryTag,  OUT record)` method?

```java
/**
		 * Set temporary state for the current message.
		 *
		 * @param correlationId the message's correlation ID, if used.
		 * @param deliveryTag the message's delivery tag.
		 */
		void setMessageIdentifiers(String correlationId, long deliveryTag) {
           // could make these checks nicer/ easier to read
			Preconditions.checkArgument(!usesCorrelationId || correlationId == null || correlationId.equals(currentCorrelationId), "Cannot use different correlation IDs for the same message");
			Preconditions.checkArgument(currentDeliveryTag == null || currentDeliveryTag == deliveryTag, "Cannot use different delivery tags for the same message");
			currentCorrelationId = correlationId;
			currentDeliveryTag = deliveryTag;
		}
```


And then if going with the individual record approach, we could do something like:

```java
@Override
		public void collect(String correlationId, long deliveryTag, OUT record) {
			setMessageIdentifiers(correlationId, deliveryTag);
			if (!autoAck) {
				if (usesCorrelationId) {
                   // could be moved into setMessageIdentifiers
					Preconditions.checkNotNull(correlationId, "RabbitMQ source was instantiated " +
						"with usesCorrelationId set to true yet we couldn't extract the correlation id from it !");
					if (!addId(correlationId)) {
						// we have already processed this message
						return;
					}
				}
				sessionIds.add(deliveryTag);
			}
			if (isEndOfStream(record)) {
				this.endOfStreamSignalled = true;
				return;
			}

			ctx.collect(record);
		}

```

Reply:
That was actually my initial approach but here is what i faced that made me split the validation logic and write the code the way i did.

Given that we also might be calling the `collect(List<T> record)` we would either have to:
* copy paste the validation logic so it runs also is checked there.
* make the `collect(List<T> record)` call the `collect(T record) method internally to add singular records but then the validation code will be called multiple times and yielding the same result.